packages/block-editor/src/hooks/block-rename-ui.js

```diff
@@ -0,0 +1,188 @@
+/**
+ * WordPress dependencies
+ */
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { addFilter } from '@wordpress/hooks';
+import { __ } from '@wordpress/i18n';
+import { getBlockSupport } from '@wordpress/blocks';
+import {
+	MenuItem,
+	__experimentalHStack as HStack,
+	__experimentalVStack as VStack,
+	Button,
+	TextControl,
+	Modal,
+} from '@wordpress/components';
+import { useState } from '@wordpress/element';
+import { useDispatch } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../store';
+import {
+	BlockSettingsMenuControls,
+	useBlockDisplayInformation,
+} from '../components';
+
+const emptyString = ( testString ) => testString?.trim()?.length === 0;
+
+function RenameModal( { blockName, originalBlockName, onClose, onSave } ) {
+	const [ editedBlockName, setEditedBlockName ] = useState( blockName );
+
+	const nameHasChanged = editedBlockName !== blockName;
+
+	const isNameValid = nameHasChanged && ! emptyString( editedBlockName );
+
+	return (
+		<Modal title={ __( 'Rename block' ) } onRequestClose={ onClose }>
+			<p>{ __( 'Choose a custom name for this block.' ) }</p>
+			<form
+				className="sidebar-navigation__rename-modal-form"
+				onSubmit={ ( e ) => {
+					e.preventDefault();
+
+					if ( ! isNameValid ) {
+						return;
+					}
+
+					onSave( editedBlockName );
+
+					// Immediate close avoids ability to hit save multiple times.
+					onClose();
+				} }
+			>
+				<VStack spacing="3">
+					<TextControl
+						__nextHasNoMarginBottom
+						value={ editedBlockName }
+						label={ __( 'Block name' ) }
+						hideLabelFromVision={ true }
+						placeholder={ __( 'Block name' ) }
+						onChange={ setEditedBlockName }
+						onBlur={ () => {
+							if ( emptyString( editedBlockName ) ) {
+								setEditedBlockName( originalBlockName );
+							}
+						} }
+					/>
+					<HStack justify="right">
+						<Button variant="tertiary" onClick={ onClose }>
+							{ __( 'Cancel' ) }
+						</Button>
+
+						<Button
+							aria-disabled={ ! isNameValid }
+							variant="primary"
+							type="submit"
+						>
+							{ __( 'Save' ) }
+						</Button>
+					</HStack>
+				</VStack>
+			</form>
+		</Modal>
+	);
+}
+
+export const withBlockRenameControl = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		const [ renamingBlock, setRenamingBlock ] = useState( false );
+		const { updateBlockAttributes } = useDispatch( blockEditorStore );
```

Comment:
The `setAttributes` method is available via `props` and stable reference.

Reply:
Oh yes! I was copying from my previous PR where we didn't have that reference, but using it would be _much_ better. Thank you for spotting.

Reply:
✅